Lib/fontParts/base/segment.py

```diff
@@ -279,7 +290,10 @@ def _get_onCurve(self):
         """
         Subclasses may override this method.
         """
-        return self.points[-1]
+        value = self.points[-1]
```

Comment:
I kindof like the invariant `segment.points[-1] == segment.onCurve`.

Since there really isn't an onCurve here (the offcurve-only quad thing really doesn't fit the segment model well at all), perhaps `segment.points[-1]` being None is also a good indicator we're having a special case here (as well as that it maintains the aforementioned invariant). It also has the advantage that code that *expects* `segment.points[-1]` to be the onCurve to be loudly broken for this case, instead of being silently broken (giving the wrong results).

Reply:
`segment.onCurve` being `None` will break less existing code then adding a `None` to `segment.points`

Reply:
My point is that it _should_ break code, because this case likely requires special handling. This point configuration wasn't supported before, so it won't break code with data it could already handle. Rephrasing: if code out there _expects_ `segment.points[-1]` to be a final on-curve (which isn't unreasonable), it will *silently* do the wrong thing with the current state of this PR.

Reply:
mmm, lots of script does a loop like:

```python
for contour in glyph:
    for segment in contour:
        for point in segment:
             print(point.x, point.y)
```

and this is silly to break them, and for script to require an extra if check 

I like that `segment.type` is `"qcurve"` now as this is only possible with quads and `segment.onCurve` is `None` as there is no on curve and `segment.offCurve` returns all off curves and looping over `[(p.x, p.y) for p in segment]` just works.

Reply:
Ok, maybe it's indeed better that way.

Reply:
@justvanrossum Are you cool with this change now? I haven't been able to review but I trust your judgement.

Reply:
I'd love to hear your judgement, too, but I think I'm ok with it. It's a tough situation that will never be completely "pretty" (it isn't pretty in `pen.qCurveTo()`, either). Let's see if I can summarize the new situation:

```python
if segment.type == "qcurve" and segment.onCurve is None:
    # on-curve-less quad blob, this must be the only segment of the parent contour
    assert segment.points == segment.offCurve
else:
    # regular segment, all points are off-curve except the last point, which is on-curve
    assert segment.points[-1] == segment.onCurve
    assert segment.points[:-1] == segment.offCurve
```
I didn't properly review the tests yet.

Reply:
This all looks good to me.